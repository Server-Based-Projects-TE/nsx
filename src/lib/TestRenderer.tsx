import type { RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'
import type { ReactElement } from 'react'
import React from 'react'
import { BrowserRouter } from 'react-router-dom'

const TestRenderer = (ui: ReactElement): RenderResult => {
  const renderResult = render(<BrowserRouter>{ui}</BrowserRouter>)
  return renderResult
}

export default TestRenderer
