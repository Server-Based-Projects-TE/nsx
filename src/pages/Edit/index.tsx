import React, { memo } from 'react'
import { useParams } from 'react-router-dom'

import { assertIsDefined } from '../../../lib/assertIsDefined'
import Button from '../../components/Button'
import BaseLayout from '../../components/Layout'
import Loading from '../../components/Loading'
import RTKQueryErrorMessages from '../../components/RTKQueryErrorMessages'
import { API } from '../../redux/API'

import useEditEffect from './useEditEffect'

const Layout: React.FC = memo(({ children }) => (
  <BaseLayout className="flex flex-col justify-start">{children}</BaseLayout>
))
Layout.displayName = 'Layout'

const Edit: React.FC = memo(() => {
  const { postId: id } = useParams() as unknown as { postId: number }
  assertIsDefined(id)

  const { data, isLoading, error } = API.endpoints.fetchPost.useQuery(id)
  const { title, body, handleEdit, handleChange } = useEditEffect(
    id,
    data,
    error
  )

  if (error) {
    return (
      <Layout>
        <RTKQueryErrorMessages error={error} />
      </Layout>
    )
  }

  if (isLoading || data === undefined) {
    return (
      <Layout>
        <Loading />
      </Layout>
    )
  }

  return (
    <Layout>
      <input
        type="text"
        className="mt-3"
        value={title}
        onChange={(e) => handleChange(e, 'title')}
        data-cy="edit-title-input"
      />
      <textarea
        className="h-60 w-full mt-3"
        value={body}
        onChange={(e) => handleChange(e, 'body')}
        data-cy="edit-body-input"
      />
      <div className="flex justify-end pt-8">
        <Button onClick={handleEdit} variant="secondary" data-cy="update-btn">
          Update
        </Button>
      </div>
    </Layout>
  )
})
Edit.displayName = 'Edit'

export default Edit
