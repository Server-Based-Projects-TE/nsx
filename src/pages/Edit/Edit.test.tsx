import { waitFor } from '@testing-library/dom'
import React from 'react'
import { Routes, Route } from 'react-router-dom'

import { sleep } from '../../../lib/sleep'
import TestRenderer from '../../lib/TestRenderer'

import Edit from './'

test('should render Edit', async () => {
  const {
    container: { firstChild },
  } = TestRenderer(
    <Routes>
      <Route path="/edit/:postId" element={<Edit />} />
    </Routes>,
    {
      initialEntries: ['/edit/3'],
    }
  )
  await waitFor(async () => {
    await sleep(500)
    expect(firstChild).toBeTruthy()
    expect(firstChild).toMatchSnapshot()
  })
})
