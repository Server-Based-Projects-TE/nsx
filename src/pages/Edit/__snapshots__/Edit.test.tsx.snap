// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit 1`] = `
<div
  class="bg-primary flex flex-col justify-between w-screen h-screen"
>
  <header
    class="w-full border-b border-gray-200"
    data-cy="header"
  >
    <div
      class="sm:flex-row sm:content-center container flex flex-col flex-wrap items-center h-16 gap-2 mx-auto my-4"
    >
      <a
        class="col-auto"
        data-cy="blog-title-top-page-link"
        href="/"
      >
        <h1
          class="sm:text-2xl text-primary text-xl font-bold"
        >
          Today I Learned
        </h1>
      </a>
      <p
        class="text-base text-secondary justify-self-start sm:pt-2.5 sm:pl-1.5"
      >
        What Today I Learned
        <span
          aria-label="note emoji"
          class="pl-1"
          role="img"
        >
          üìù
        </span>
      </p>
      <div
        class="sm:flex-grow sm:justify-end flex space-x-2"
      >
        <a
          href="/about"
        >
          <div
            class="hover:text-gray-300 text-secondary text-lg"
          >
            about
          </div>
        </a>
        <label
          class="sr-only"
          id="headlessui-listbox-label-1"
        >
          Theme
        </label>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          aria-labelledby="headlessui-listbox-label-1 headlessui-listbox-button-undefined"
          id="headlessui-listbox-button-2"
          type="button"
        >
          <span
            class="dark:hidden"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
            >
              <path
                class="stroke-slate-400 dark:stroke-slate-500"
                d="M15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0Z"
              />
              <path
                class="stroke-slate-400 dark:stroke-slate-500"
                d="M12 4v1M17.66 6.344l-.828.828M20.005 12.004h-1M17.66 17.664l-.828-.828M12 20.01V19M6.34 17.664l.835-.836M3.995 12.004h1.01M6 6l.835.836"
              />
            </svg>
          </span>
          <span
            class="dark:inline hidden"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              viewBox="0 0 24 24"
            >
              <path
                class="fill-transparent"
                clip-rule="evenodd"
                d="M17.715 15.15A6.5 6.5 0 0 1 9 6.035C6.106 6.922 4 9.645 4 12.867c0 3.94 3.153 7.136 7.042 7.136 3.101 0 5.734-2.032 6.673-4.853Z"
                fill-rule="evenodd"
              />
              <path
                class="fill-slate-400 dark:fill-slate-500"
                d="m17.715 15.15.95.316a1 1 0 0 0-1.445-1.185l.495.869ZM9 6.035l.846.534a1 1 0 0 0-1.14-1.49L9 6.035Zm8.221 8.246a5.47 5.47 0 0 1-2.72.718v2a7.47 7.47 0 0 0 3.71-.98l-.99-1.738Zm-2.72.718A5.5 5.5 0 0 1 9 9.5H7a7.5 7.5 0 0 0 7.5 7.5v-2ZM9 9.5c0-1.079.31-2.082.845-2.93L8.153 5.5A7.47 7.47 0 0 0 7 9.5h2Zm-4 3.368C5 10.089 6.815 7.75 9.292 6.99L8.706 5.08C5.397 6.094 3 9.201 3 12.867h2Zm6.042 6.136C7.718 19.003 5 16.268 5 12.867H3c0 4.48 3.588 8.136 8.042 8.136v-2Zm5.725-4.17c-.81 2.433-3.074 4.17-5.725 4.17v2c3.552 0 6.553-2.327 7.622-5.537l-1.897-.632Z"
              />
              <path
                class="fill-slate-400 dark:fill-slate-500"
                clip-rule="evenodd"
                d="M17 3a1 1 0 0 1 1 1 2 2 0 0 0 2 2 1 1 0 1 1 0 2 2 2 0 0 0-2 2 1 1 0 1 1-2 0 2 2 0 0 0-2-2 1 1 0 1 1 0-2 2 2 0 0 0 2-2 1 1 0 0 1 1-1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
        <a
          href="https://github.com/laststance/nsx"
          rel="noreferrer"
          target="_blank"
        >
          <svg
            class="inline w-6 h-6 text-gray-500"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <title>
              Github
            </title>
            <path
              clip-rule="evenodd"
              d="M12 2C6.477 2 2 6.484 2 12.017c0 4.425 2.865 8.18 6.839 9.504.5.092.682-.217.682-.483 0-.237-.008-.868-.013-1.703-2.782.605-3.369-1.343-3.369-1.343-.454-1.158-1.11-1.466-1.11-1.466-.908-.62.069-.608.069-.608 1.003.07 1.531 1.032 1.531 1.032.892 1.53 2.341 1.088 2.91.832.092-.647.35-1.088.636-1.338-2.22-.253-4.555-1.113-4.555-4.951 0-1.093.39-1.988 1.029-2.688-.103-.253-.446-1.272.098-2.65 0 0 .84-.27 2.75 1.026A9.564 9.564 0 0112 6.844c.85.004 1.705.115 2.504.337 1.909-1.296 2.747-1.027 2.747-1.027.546 1.379.202 2.398.1 2.651.64.7 1.028 1.595 1.028 2.688 0 3.848-2.339 4.695-4.566 4.943.359.309.678.92.678 1.855 0 1.338-.012 2.419-.012 2.747 0 .268.18.58.688.482A10.019 10.019 0 0022 12.017C22 6.484 17.522 2 12 2z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </div>
    </div>
  </header>
  <main
    class="container mx-auto flex-grow px-4 py-4 flex flex-col justify-start"
  >
    <input
      class="mt-3"
      data-cy="edit-title-input"
      type="text"
      value="Loading..."
    />
    <textarea
      class="h-60 w-full mt-3"
      data-cy="edit-body-input"
    >
      Loading...
    </textarea>
    <div
      class="flex justify-end pt-8"
    >
      <button
        class="flex justify-center items-center shadow rounded disabled:opacity-70 disabled:cursor-not-allowed focus:outline-none font-semibold bg-blue-500 hover:bg-blue-600 text-white py-2 px-6 text-md"
        data-cy="update-btn"
        type="button"
      >
        <span
          class="mx-2"
        >
          Update
        </span>
      </button>
    </div>
  </main>
  <footer
    class="w-full"
  >
    <div
      class="flex flex-col items-center w-full px-6 border-t border-gray-200"
    >
      <div
        class="sm:w-2/3 py-6 text-center"
      >
        <p
          class="text-sm text-gray-600"
        >
          ¬© 2022 by Ryota Murakami
        </p>
      </div>
    </div>
  </footer>
</div>
`;
